library/core/src/macros/mod.rs

```diff
@@ -548,13 +578,30 @@ macro_rules! write {
 #[macro_export]
 #[stable(feature = "rust1", since = "1.0.0")]
 #[cfg_attr(not(test), rustc_diagnostic_item = "writeln_macro")]
-#[allow_internal_unstable(format_args_nl)]
+#[allow_internal_unstable(autoref, format_args_nl)]
 macro_rules! writeln {
     ($dst:expr $(,)?) => {
         $crate::write!($dst, "\n")
     };
+
+    (
+        $dst_place:ident $(. $field_or_idx:tt)* ,
+        $($arg:tt)*
+    ) => ({
+        let result =
+            $dst_place $(. $field_or_idx)*
+                .write_fmt($crate::format_args_nl!($($arg)*))
+        ;
+        result
+    });
+
     ($dst:expr, $($arg:tt)*) => {
-        $dst.write_fmt($crate::format_args_nl!($($arg)*))
+        match $crate::ops::autoref::autoref_mut!($dst) {
+            mut _dst => {
```

Comment:
`_dst` is guaranteed to be `&mut _`, doesn't that mean the `mut` here is redundant?
(Also, is that why you had to add a `_` prefix? Doesn't really seem needed otherwise AFAICT)

Reply:
triage: 
@danielhenrymantilla ☝️ - can you address this? thanks

Reply:
Ah, yes, sorry, I had completely forgotten about this comment 😅. So the thing is that sometimes you do need a nested `&mut`, such as for the `&mut [u8] : Write` case:

```rs
let mut buf: &mut [u8] = &mut [][..];
buf.write_fmt(...) // resolves to `Write::write_fmt(&mut buf : &mut &mut [u8], ...)`
// vs
match autoref!(buf) { // here, the identity function: `match identity::<&mut [u8]>(buf) {`
    _buf => _buf.write_fmt(...), // we need the *nested* `&mut`, and for that, we need `_buf` to be `mut`
```

  - [Demo](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=31732ab56b5950ee3513e7df8c2fcc84)

So the `mut` is necessary, and, from there, either `#[allow(unused_mut)]` or a `_`-prefixed binding name is needed (ideally this macro-generated stuff would not lint altogether, although I guess it's not possible for `core` / the crate where the macro is defined). I went for `_`-prefixed as a personal choice since it yields slightly more lightweight code, but I can change it to the `allow`.